enterprise_gateway/itests/src/main.py

```diff
@@ -14,6 +14,7 @@ def parse_arg():
     parser.add_argument('--continue_when_error', default=True)
     parser.add_argument('--host', default='localhost:8888')
     parser.add_argument('--username', default='root')
+    parser.add_argument('--impersonation_username', default=None)
```

Comment:
I prefer we not introduce another `username` parameter.  Can we just use `username` and change `--impersonation_username` be a boolean flag that indicates that impersonation be enforced?  (e.g., `--enforce_impersonation`)
`username` will ALWAYS equal `impersonation_username`, so the latter is redundant.

Reply:
Not sure how `usename` is used, but if it's purely there to set into `KERNEL_USERNAME` or not generally used, then we should probably default it to `None` and make the setting of `KERNEL_USERNAME` conditional to mimic users not setting `KERNEL_USERNAME`.

Reply:
The `username` parameter is only used as `KERNEL_USERNAME` when a new kernel is to be created:
https://github.com/jupyter-incubator/enterprise_gateway/blob/master/enterprise_gateway/itests/src/elyra_client.py#L31

And yes I'd better change the argument's name as `enforce_impersonation`, a boolean flag. When it is set as `True`, the `username` and the test output username should be compared to verify if it is a valid result of impersonation.

Will make the change.

Reply:
I would recommend setting the default value as a non-None value e.g. `username` as `root` or `elyra`.

Based on my observation, it is ok to open new kernels in client mode with `None` as the `KERNEL_USERNAME`/`username`. But for all kernels in cluster mode, there would be an error like this:

```
[D 2017-12-08 11:28:15.887 EnterpriseGatewayApp] BaseProcessProxy.launch_process() env: {u'EG_ALTERNATE_SIGINT': u'SIGUSR2', u'__TOREE_SPARK_OPTS__': u'--master yarn --deploy-mode cluster --name ${KERNEL_ID:-ERROR__NO__KERNEL_ID} --conf spark.yarn.submit.waitAppCompletion=false --conf spark.yarn.am.waitTime=1d', u'SPARK_YARN_USER_ENV': u'TOREE_ALTERNATE_SIGINT=USR2', 'KERNEL_ID': u'4def8b26-1045-4987-9f4f-edabda6fc199', u'SPARK_HOME': u'/usr/hdp/current/spark2-client', 'KERNEL_GATEWAY': '1', u'TOREE_ALTERNATE_SIGINT': u'USR2', u'__TOREE_OPTS__': u'', u'KERNEL_USERNAME': None, u'DEFAULT_INTERPRETER': u'Scala', 'PATH': '/opt/anaconda2/bin:/usr/lib64/qt-3.3/bin:/opt/anaconda2/bin:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/elyra/.local/bin:/home/elyra/bin', u'LAUNCH_OPTS': u''}
[E 2017-12-08 11:28:15.896 EnterpriseGatewayApp] Failed to run command:
    [u'/usr/local/share/jupyter/kernels/spark_scala_yarn_cluster/bin/run.sh', u'--profile', u'/home/elyra/.local/share/jupyter/runtime/kernel-4def8b26-1045-4987-9f4f-edabda6fc199.json', u'--RemoteProcessProxy.response-address', u'172.16.155.185:44558']
        PATH='/opt/anaconda2/bin:/usr/lib64/qt-3.3/bin:/opt/anaconda2/bin:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/elyra/.local/bin:/home/elyra/bin'
        with kwargs:
    {'cwd': None, 'stdin': -1, 'preexec_fn': <function <lambda> at 0x7f6dd01c2668>, 'stderr': None, 'stdout': None}
    
[E 2017-12-08 11:28:15.897 EnterpriseGatewayApp] Unhandled error in API request
    Traceback (most recent call last):
      File "/opt/anaconda2/lib/python2.7/site-packages/notebook/base/handlers.py", line 540, in wrapper
        result = yield gen.maybe_future(method(self, *args, **kwargs))
      File "/opt/anaconda2/lib/python2.7/site-packages/tornado/gen.py", line 1055, in run
        value = future.result()
      File "/opt/anaconda2/lib/python2.7/site-packages/tornado/concurrent.py", line 238, in result
        raise_exc_info(self._exc_info)
      File "/opt/anaconda2/lib/python2.7/site-packages/tornado/gen.py", line 1063, in run
        yielded = self.gen.throw(*exc_info)
      File "/opt/anaconda2/lib/python2.7/site-packages/notebook/services/kernels/handlers.py", line 49, in post
        kernel_id = yield gen.maybe_future(km.start_kernel(kernel_name=model['name']))
      File "/opt/anaconda2/lib/python2.7/site-packages/tornado/gen.py", line 1055, in run
        value = future.result()
      File "/opt/anaconda2/lib/python2.7/site-packages/tornado/concurrent.py", line 238, in result
        raise_exc_info(self._exc_info)
      File "/opt/anaconda2/lib/python2.7/site-packages/tornado/gen.py", line 1063, in run
        yielded = self.gen.throw(*exc_info)
      File "/opt/anaconda2/lib/python2.7/site-packages/enterprise_gateway/services/kernels/remotemanager.py", line 28, in start_kernel
        kernel_id = yield gen.maybe_future(super(RemoteMappingKernelManager, self).start_kernel(*args, **kwargs))
      File "/opt/anaconda2/lib/python2.7/site-packages/tornado/gen.py", line 1055, in run
        value = future.result()
      File "/opt/anaconda2/lib/python2.7/site-packages/tornado/concurrent.py", line 238, in result
        raise_exc_info(self._exc_info)
      File "/opt/anaconda2/lib/python2.7/site-packages/tornado/gen.py", line 1063, in run
        yielded = self.gen.throw(*exc_info)
      File "/opt/anaconda2/lib/python2.7/site-packages/kernel_gateway/services/kernels/manager.py", line 81, in start_kernel
        kernel_id = yield gen.maybe_future(super(SeedingMappingKernelManager, self).start_kernel(*args, **kwargs))
      File "/opt/anaconda2/lib/python2.7/site-packages/tornado/gen.py", line 1055, in run
        value = future.result()
      File "/opt/anaconda2/lib/python2.7/site-packages/tornado/concurrent.py", line 238, in result
        raise_exc_info(self._exc_info)
      File "/opt/anaconda2/lib/python2.7/site-packages/tornado/gen.py", line 307, in wrapper
        yielded = next(result)
      File "/opt/anaconda2/lib/python2.7/site-packages/notebook/services/kernels/kernelmanager.py", line 123, in start_kernel
        super(MappingKernelManager, self).start_kernel(**kwargs)
      File "/opt/anaconda2/lib/python2.7/site-packages/jupyter_client/multikernelmanager.py", line 110, in start_kernel
        km.start_kernel(**kwargs)
      File "/opt/anaconda2/lib/python2.7/site-packages/enterprise_gateway/services/kernels/remotemanager.py", line 100, in start_kernel
        return super(RemoteKernelManager, self).start_kernel(**kw)
      File "/opt/anaconda2/lib/python2.7/site-packages/jupyter_client/manager.py", line 257, in start_kernel
        **kw)
      File "/opt/anaconda2/lib/python2.7/site-packages/enterprise_gateway/services/kernels/remotemanager.py", line 126, in _launch_kernel
        return self.process_proxy.launch_process(kernel_cmd, **kw)
      File "/opt/anaconda2/lib/python2.7/site-packages/enterprise_gateway/services/processproxies/yarn.py", line 57, in launch_process
        self.local_proc = launch_kernel(kernel_cmd, **kw)
      File "/opt/anaconda2/lib/python2.7/site-packages/jupyter_client/launcher.py", line 128, in launch_kernel
        proc = Popen(cmd, **kwargs)
      File "/opt/anaconda2/lib/python2.7/subprocess.py", line 390, in __init__
        errread, errwrite)
      File "/opt/anaconda2/lib/python2.7/subprocess.py", line 1024, in _execute_child
        raise child_exception
    TypeError: execve() arg 3 contains a non-string value
[E 171208 11:28:15 web:2063] 500 POST /api/kernels (9.98.25.12) 21.20ms
```

Reply:
@liukun1016 - this issue looks familiar and something I believe I addressed in a recent change.  Are you sure you're running against the latest version?  At least the version that adds support for impersonation and authorization?